File: analysis/figure_scripts/phi.cpp
// ROOT stuff
#include <TFile.h>
#include <TTree.h>
#include <TChain.h>
#include <ROOT/RDataFrame.hxx>
#include <ROOT/RDFHelpers.hxx>
#include <TStyle.h>
#include <TROOT.h>
#include <TCanvas.h>

// other stuff
#include <filesystem>
#include <boost/format.hpp>
#include <iostream>

// my stuff
#include "../plot_style.cpp"

using namespace std;
using namespace ROOT;
using boost::format;

int main(int argc, char *argv[]) {
    // all but the last argument are input files
    TChain chain("tree");
    for (int i = 1; i < argc-1; i++) {
        chain.Add(argv[i]);
    }    
    // final argument is the destination
    string dest = argv[argc-1]; 
    dest += "phi.pdf";

    // prepare the dataframe
    ROOT::RDF::RNode df = RDataFrame(chain);
    df = df.Define("E_sum","E_cm[0]+E_cm[1]+E_cm[2]");

    // set the axes
    double x_axis[] = {200, 0, M_PI/3};

    // define the necessary variables
    df = df.Filter("p_tot<35e3")
           .Filter("abs(deltaE)<200")
           .Define("E2","min(E_cm[0],min(E_cm[1],E_cm[2]))")
           .Define("E1","max(min(E_cm[0],E_cm[1]), min(max(E_cm[0],E_cm[1]),E_cm[2]))")\
           .Define("E0","max(E_cm[0],max(E_cm[1],E_cm[2]))")
           .Define("X","((E1-E2)/E_sum)*sqrt(3)")
           .Define("Y","3*E0/E_sum-1.0")
           .Filter("pow(X,2)+pow(Y,2)<1.0")
           .Filter("Y<0.93")
           .Filter("E2>250");

    //*** PLOT ***//
    setup_style();

    TCanvas* canvas = new TCanvas("phi", "phi", 400, 400);
    TH1D hist = df.Define("x", "atan2(X,Y)")
                  .Histo1D({"h1", "phi", int(x_axis[0]), x_axis[1], x_axis[2]}, "x").GetValue();

    hist.GetXaxis()->SetTitle("\\phi");
    hist.GetYaxis()->SetTitle("Count");
    hist.Draw("colz");

    canvas->SetRightMargin(0.15);
    canvas->SaveAs(dest.c_str());
}

File: analysis/figure_scripts/phi.py
import matplotlib.pyplot as plt
import sys
from ROOT import TChain
from dataframe import *

chain = TChain('tree')
for f in input().split(): chain.AddFile(f)
fname = sys.argv[1]+"/"+sys.argv[0].split('/')[2].split('.')[0]

r = dataframe(chain)
r = r.define("esum","E_cm[0]+E_cm[1]+E_cm[2]")
        #.define("X","((E_cm[0]+2*E_cm[1])/esum-1)/sqrt(3)")\
        #.define("Y","E_cm[0]/esum-1.0/3.0")\
        #.define("phi","fmod(atan2(Y,X)+TMath::Pi(),2*TMath::Pi()/3)")\
        #.filter("pow(X,2)+pow(Y,2)<1.0/9.0")\
lim = (0,3.1416/3)
r = r.filter("p_tot<35e3")\
    .filter("abs(deltaE)<200")\
    .define("E2","min(E_cm[0],min(E_cm[1],E_cm[2]))")\
    .define("E1","max(min(E_cm[0],E_cm[1]), min(max(E_cm[0],E_cm[1]),E_cm[2]))")\
    .define("E0","max(E_cm[0],max(E_cm[1],E_cm[2]))")\
    .define("X","((E1-E2)/esum)*sqrt(3)")\
    .define("Y","3*E0/esum-1.0")\
    .filter("pow(X,2)+pow(Y,2)<1.0")\
    .filter("Y<0.93")\
    .filter("E2>250")

ghist(r,"atan2(X,Y)",300,lim,fname)
#if "wU" in r.getColumnNames():
    #r.hist("atan2(X,Y)",bins=300,range=(lim),export=(fname,"f1"),weights="wU*f[0][0]")
    #r.hist("atan2(X,Y)",bins=300,range=(lim),export=(fname,"f2"),weights="wU*f[0][1]")
    #r.hist("atan2(X,Y)",bins=300,range=(lim),export=(fname,"im"),weights="wU*f[0][3]")
    #r.hist("atan2(X,Y)",bins=300,range=(lim),export=(fname,"re"),weights="wU*f[0][2]")

#if "data" in fname:
    #r.hist("atan2(X,Y)",bins=300,range=(lim),export=(fname,"count"))
#else:
    #r.hist("atan2(X,Y)",bins=300,range=(lim),export=(fname,"count"),weights="w0*w1*w2")
#plt.clf()
#if "wU" in r.getColumnNames():
    #r.hist("atan2(X,Y)",bins=300,range=(lim),weights="wU*(0.22*f[0][0]+(1-0.22)*f[0][1]+2*sqrt(0.22*(1-0.22))*(f[0][2]*cos(2*3.1415*0.69)+f[0][3]*sin(2*3.1415*0.69)))*w0*w1*w2")
#else:
    #r.hist("atan2(X,Y)",bins=300,range=(lim))

plt.title(sys.argv[1])
plt.xlabel("$\\phi$")
plt.ylabel("Count")

plt.savefig(fname+".pdf",dpi=600)
